### ISSUE TITLE
Change Drivers Cabin

### ISSUE THREAD

**Issue by DasCapschen** [Libre-TrainSim/Libre-TrainSim#130]
Depends on #95.

Once driving backwards works, we should allow the player to change the drivers cabin in order to change the direction the train is driving.  
This would basically just mean inverting the `reverser` setting (backwards becomes forwards) and facing the player in the opposite direction.  
We should **not** flip the train model around. Instead only allow changing cabin if there is a cabin at both ends of the train. (which is currently the case for the JFR)
This would require adding a second "cabin" to the train, or perhaps moving only the cabin to a new location / rotating it by 180 degrees.

**Comment by HaSa1002** [comment]
I lobby for combined wagons and cabins, meaning a cabin is just child node of a wagon and has a script attached that is inheriting `Cabin`. We can query the children in the wagon script on ready with
```gdscript
for child in get_children():
	if child is Cabin:
		# do something with cabin like
		cabines.push_back(child)
```
or simply add an export
```gdscript
export(Array, NodePath) cabins := []
```

**Comment by DasCapschen** [comment]
I agree with this. Not sure if we necessarily need a Cabin script (does the cabin do anything special?), but I agree the cabin should just be a normal Wagon.
I already have some changes (although still limited to 1 cabin) in one of my branches: https://github.com/DasCapschen/Libre-TrainSim/tree/refactor_player which basically makes cabin mode optional, so the train can function without a "cabin", to allow for trains where the cabin is integrated into the normal wagon model.

**Comment by HaSa1002** [comment]
I would refactor all cabin logic out of wagon and train into cabin

**Comment by DasCapschen** [comment]
Well currently the only "logic" is hiding the wagon models and showing the cabin model. It literally does nothing else. So that's why I'm not sure if we need a separate cabin script.

**Comment by HaSa1002** [comment]
I mean stuff like acting controls, showing displays and thelike. I think it is currently specific scripting.

**Comment by DasCapschen** [comment]
Specific Scripting is currently stuff like animating the controls in the cabin and displays, yes. This could indeed be in a sort of "cabin" script, that is then attached to the wagon instance which acts as the cabin, good idea :+1:

**Comment by Jean28518** [comment]
Why would you refactor the cabin out of wagon.gd?
Wagon and Cabin are sharing very much functionality, e.g. the sound system needs the wagon script.... 

I also don't recommend to move out some controls or displays out of specific scripting, because every train works different with these components...

I recommend to not switch and rotate the same cabin to the end of the train, that could mess up, if we want to use different wagontypes at beginning and front.

**Comment by DasCapschen** [comment]
I think the idea was to have `Wagon` be a class, and for `Cabin` to extend `Wagon`. 
Then you can put all the Cabin-Only code in it's own file, instead of having it mixed in with the Wagons. 

And then I think the idea was to have each train, if it needs to modify how the cabin works, extend `Cabin` to its own custom script.
We could still have a `SpecificScripting` node for special non-cabin train behaviour. 

I think the main idea is to have things more clearly separated.

**Comment by DasCapschen** [comment]
Oh and yes, we don't want to re-use the same cabin by just rotating it. The train creator would have to setup the train so that there are 2 cabins to select from.

**Comment by Jean28518** [comment]
Okay, good ideas!
